Simplified/AxisService/NYPLAxisContentDownloader.swift

```diff
@@ -0,0 +1,153 @@
+//
+//  NYPLAxisContentDownloader.swift
+//  Simplified
+//
+//  Created by Raman Singh on 2021-04-22.
+//  Copyright © 2021 NYPL Labs. All rights reserved.
+//
+
+import Foundation
+
+protocol NYPLAxisContentDownloading {
+  func downloadItem(from url: URL,
+                    _ completion: @escaping (Result<Data, Error>) -> Void)
+}
+
+
+/// A content downloader which stops downloading content upon receiving a failure.
+class NYPLAxisContentDownloader: NYPLAxisContentDownloading {
+  
+  private let tasks: ThreadSafeDictionaryContainer<URL, URLSessionDataTask>
+  private let handlers: ThreadSafeDictionaryContainer<URL, (Result<Data, Error>) -> Void>
+  private let numberOfAttempts: ThreadSafeDictionaryContainer<URL, Int>
+  private let errored: ThreadSafeValueContainer<Bool>
+  let networkExecutor: NYPLAxisNetworkExecuting
+  
+  init(networkExecuting: NYPLAxisNetworkExecuting) {
+    self.tasks = ThreadSafeDictionaryContainer()
+    self.handlers = ThreadSafeDictionaryContainer()
+    self.numberOfAttempts = ThreadSafeDictionaryContainer()
+    self.errored = ThreadSafeValueContainer(value: false)
+    self.networkExecutor = networkExecuting
+  }
+  
+  func downloadItem(from url: URL,
+                    _ completion: @escaping (Result<Data, Error>) -> Void) {
+    
+    // We dont want to download anything if it's already being downloaded or if
+    // we experienced an error downloading something before.
+    guard !errored.value, tasks[url] == nil, handlers[url] == nil else {
+      return
+    }
+    
+    numberOfAttempts[url] = (self.numberOfAttempts[url] ?? 0) + 1
+    handlers[url] = completion
+    tasks[url] = self.getTask(for: url)
+  }
+  
+  private func getTask(for url: URL) -> URLSessionDataTask {
+    
+    let request = URLRequest(url: url,
+                             cachePolicy: .reloadIgnoringLocalCacheData,
+                             timeoutInterval: networkExecutor.requestTimeout)
+    
+    let task = networkExecutor.GET(request) { [weak self] (result) in
+      guard let self = self else { return }
+      switch result {
+      case .success(let data):
+        self.handlers[url]?(.success(data))
+        self.tasks[url] = nil
+        self.handlers[url] = nil
+      case .failure(let error):
+        // In poor network conditions, the download might fail on first or
+        // second attempt. We retry 2 more times to make sure we don't
+        // prematurely stop the download process.
+        if
```

Comment:
@ErnestFan @ettore In my experiments with downloading book content while having poor network conditions, there were 130 items that failed to download in the first attempt, and 56 of those items failed to download the second time, but all those 56 items successfully downloaded on the 3rd attempt.

So, to make sure we're doing everything we can to ensure a book download succeeds in all network conditions, should I up the retry count to 5?

Reply:
I am not sure if we should up the retry count or let the user hit the download button again when they feel ready gives better user experience. Let's see what's Ettore's input.

Reply:
this is a tough one, I would say that with poor network conditions all bets are off. With more automatic retries the user might be seeing a spinner for a long time and then it might _still_ fail, and that would be even worse. 

I'd like to understand a few things first: 
- Is there one call to `getTask(for:)` for each of the 130 items? 
- And are those 130 items for just one book?

I'm wondering if these failures are caused because we are flooding the network with too many requests at the same time, but I don't know exactly how these requests are submitted and how Axis works. (In my ignorance I was expecting to download a zip file or something like that.)

Reply:
@ettore This is on me. I should had written a brief overview of how Axis works at the top of **AxisService.swift** file. Here's how it works:

1. When we send a download request to server, we get a file with isbn & book_vault_uuid key. (`NYPLMyBooksDownloadCenter::URLSession:session:downloadTask:didFinishDownloadingToURL`)
2. If the book's management rights are the ones with axis, we initiate NYPLAxisService and call the `NYPLAxisService::fulfillAxisLicense:downloadTask` method.
3. Using isbn & book_vault_uuid from the file, we create a url to download a license file for the axis book. (NYPLAxisService::downloadLicense).
4. If the license download is successful, we download two files **Container.xml** & **Encryption.xml**) from Axis using the isbn key. (`NYPLAxisService::downloadContainer` & `NYPLAxisService::downloadEncryption`)
5. From the **container.xml** file, we extract the url for a **package.opf** file and download the **package.opf** file (`NYPLAxisService::downloadPackage`).
6. From the **package.opf** file (which is more of an xml really), we extract urls for all the required files for the book (e.g. chapter1.xtml, cover.html, fonts, images, styles etc.). We pass those urls in a for loop and tell the downloader to download all the files. (`NYPLAxisService::downloadAssetsFromPackage`)
7. When all the files are downloaded, we consider this as book download success.

Reply:
So to answer your questions,
_Is there one call to getTask(for:) for each of the 130 items?_
We are storing tasks in a dictionary using the associated url as the key. We achieve two things by doing this.
1. We don't download from the same url twice (if for any reason the package.opf file contains two files with the exact same url). We do this by making sure tasks[url] is nil before initiating the download in `NYPLAxisContentDownloader::downloadItem:from completion`.
2. We can cancel all the pending downloads if a file fails to download (after 3 tries).

_And are those 130 items for just one book?_
Yes. An Axis book folder structure looks like this
fonts (directory)
images (directory)
styles (directory)
toc.ncx
xhtml (directory)

The file size can vary depending on the book.

Reply:
thank you for the detailed explanation, this helps a lot. It would be great if you can add this write-up to the code base for future reference.

going back to the original doubt...I'm not opposed to the retry strategy, to me 3 attempts seem reasonable, I was just wondering if we are inducing that by flooding the network. Hundreds of requests could do that -- although come to think of it, I believe NYPLNetworkExecutor has already infrastructure to cap to 4 at a time max. Perhaps verify that, my memory is faulty... 😭

More than everything it would be good to double check the network behavior with a proxy just to be sure we are actually issuing 4 - 8 requests at a time. (I use Charles, it has a free tier, but it costs only $50 and it's worth every penny... there are many others though)

Reply:
That's a good point. Just checked and even though by the default value is 6 in macOS and 4 in iOS, we're using a URLSessionConfiguration created in NYPLCaching which has this value amped up to maximum 8 connections per host.

`NYPLCaching::makeURLSessionConfiguration:caching:requestTimeout`